Source/MaestroBurner.cpp

```diff
@@ -94,6 +94,169 @@ void Maestro::Burner(const Vector<MultiFab>& s_in, Vector<MultiFab>& s_out,
                           : rho_Hext[lev].array(mfi);
             const Array4<const int> mask_arr = mask.array(mfi);
 
+            // If save_react_data, we add perturbations to the mass fractions to
+            // get more varied data for training ML model
+            if (save_react_data) {
+                ParallelForRNG(tileBox, [=] AMREX_GPU_DEVICE(int i, int j, int k,
+                                                             amrex::RandomEngine const& engine) noexcept
+                {
+                    if (use_mask && mask_arr(i, j, k))
+                        return;  // cell is covered by finer cells
+
+                    auto rho = s_in_arr(i, j, k, Rho);
+                    Real x_in[NumSpec];
+                    for (int n = 0; n < NumSpec; ++n) {
+                        x_in[n] = s_in_arr(i, j, k, FirstSpec + n) / rho;
+                    }
+
+#if NAUX_NET > 0
+                    Real aux_in[NumAux];
+                    for (int n = 0; n < NumAux; ++n) {
+                        aux_in[n] = s_in_arr(i, j, k, FirstAux + n) / rho;
+                    }
+#endif
+
+		    Real T_in = 0.0;
+		    if (drive_initial_convection) {
+			if (!spherical) {
+			    auto r = (AMREX_SPACEDIM == 2) ? j : k;
+			    T_in = tempbar_init_arr(lev, r);
+			} else {
+			    T_in = tempbar_cart_arr(i, j, k);
+			}
+		    } else {
+			T_in = s_in_arr(i, j, k, Temp);
+		    }
+
+		    Real x_test =
+			(ispec_threshold > 0) ? x_in[ispec_threshold] : 0.0;
+
+		    // Perturb 75 percent of the data
+		    if (amrex::Random(engine) < 0.75) {
+			// Three main species are C12(1), O16(2), Mg24(4)
+			// Loop over rest of the species
+			for (int n = 0; n < NumSpec; ++n) {
+			    // Do not perturb the main species
+			    if (n != 1 && n != 2 && n !=4) {
+				// Set a random perturbation based on log(X_k)
+				Real X_log = log(x_in[n]);
```

Comment:
use std::log10() to make things easier to follow?

Reply:
I think exp() is faster than pow() though? I've also been using exp() during training.

Reply:
Not sure, but I don't think this will affect overall runtime much one way or another.

Reply:
Okay, well it is really only a factor of log(10) difference, so it may not matter much in terms of runtime. I'll change it.